src/number_theory/legendre_symbol/add_character.lean

```diff
@@ -0,0 +1,323 @@
+abbreviation add_char : Type (max u v) := (multiplicative R) →* R'
+
+end add_char_def
+
+
+section additive
+
+namespace add_char
+
+open multiplicative
+
+universes u v
+
+-- The domain of our additive characters
+variables {R : Type u} [comm_ring R]
+-- The target
+variables {R' : Type v} [comm_ring R']
+
+/-- The trivial additive character; it sends everything to `1`. -/
+@[simps] protected
+def trivial : add_char R R' :=
+{ to_fun := λ x, 1,
+  map_one' := rfl,
+  map_mul' := λ x y, (mul_one 1).symm }
+
+/-- An additive character is *nontrivial* if it takes a value `≠ 1`. -/
+def is_nontrivial (ψ : add_char R R') : Prop := ∃ (a : R), ψ a ≠ 1
+
+/-- An additive character is nontrivial iff it is not the trivial character. -/
+lemma is_nontrivial_iff_ne_trivial (ψ : add_char R R') :
+  is_nontrivial ψ ↔ ψ ≠ add_char.trivial :=
+begin
+  refine not_forall.symm.trans (iff.not _),
+  rw fun_like.ext_iff,
+  refl,
+end
+
+/-- Define the multiplicative shift of an additive character.
+This satisfies `mul_shift ψ a x = ψ (a * x)`. -/
+@[simps]
+def mul_shift (ψ : add_char R R') (a : R) : add_char R R' :=
+ψ.comp (add_monoid_hom.mul_left a).to_multiplicative
+
+lemma mul_shift_spec {ψ : add_char R R'} {a : R} {x : multiplicative R} :
+  mul_shift ψ a x = ψ (of_add (a * x)) := rfl
+
+/-- If `n` is a natural number, then `mul_shift ψ n x = (ψ x) ^ n`. -/
+lemma mul_shift_spec' (ψ : add_char R R') (n : ℕ) (x : R) : mul_shift ψ n x = (ψ x) ^ n :=
+begin
+  rw [mul_shift_spec, ← nsmul_eq_mul, of_add_nsmul, map_pow],
+  refl,
+end
+
+/-- The product of `mul_shift ψ a` and `mul_shift ψ b` is `mul_shift ψ (a + b)`. -/
+lemma mul_shift_mul (ψ : add_char R R') (a b : R) :
+  mul_shift ψ a * mul_shift ψ b = mul_shift ψ (a + b) :=
+begin
+  ext,
+  simp only [monoid_hom.mul_apply, mul_shift_spec, add_mul, of_add_add, ψ.map_mul],
+end
+
+/-- `mul_shift ψ 0` is the trivial character. -/
+@[simp]
+lemma mul_shift_zero (ψ : add_char R R') : mul_shift ψ 0 = add_char.trivial :=
+begin
+  ext,
+  simp only [mul_shift_apply, zero_mul, of_add_zero, map_one, trivial_apply],
+end
+
+/-- An additive character is *primitive* iff all its multiplicative shifts by nonzero
+elements are nontrivial. -/
+def is_primitive (ψ : add_char R R') : Prop :=
+∀ (a : R), a ≠ 0 → is_nontrivial (mul_shift ψ a)
+
+/-- The map associating to `a : R` the multiplicative shift of `ψ` by `a`
+is injective when `ψ` is primitive. -/
+lemma to_mul_shift_inj_of_is_primitive {ψ : add_char R R'} (hψ : is_primitive ψ) :
+  function.injective ψ.mul_shift :=
+begin
+  intros a b h,
+  have h₁ : mul_shift ψ a * mul_shift ψ (- b) = mul_shift ψ b * mul_shift ψ (- b) :=
+    congr_fun (congr_arg has_mul.mul h) (mul_shift ψ (- b)),
+  simp only [mul_shift_mul, mul_shift_zero, add_right_neg] at h₁,
+  have h₂ := hψ (a + (- b)),
+  rw [h₁, is_nontrivial_iff_ne_trivial, ← sub_eq_add_neg, ne.def add_char.trivial,
+      ne_self_iff_false, imp_false, not_not] at h₂,
+  exact eq_of_sub_eq_zero h₂,
```

Comment:
```suggestion
  intros a b h,
  apply_fun (λ x, x * mul_shift ψ (-b)) at h,
  simp only [mul_shift_mul, mul_shift_zero, add_right_neg] at h,
  have h₂ := hψ (a + (- b)),
  rw [h, is_nontrivial_iff_ne_trivial, ← sub_eq_add_neg, sub_ne_zero] at h₂,
  exact not_not.mp (λ h, h₂ h rfl),
```

Using `apply_fun` saves you from having to make h\1.

Reply:
OK.

Reply:
OK.